Table Editor & Columns Fusion

buthanoid opened the pull request:
Table with fusion of Output/Input/HardCoded columns. Table Editor to filter columns. Pretty rendering for SUCCESS and RATING.

bourgesl commented on src/main/java/fr/jmmc/oimaging/gui/SuccessCell.java:
```diff
@@ -37,7 +37,7 @@ public SuccessCell() {
     @Override
     public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
-        success = (boolean) value;
+        if (value != null) success = (boolean) value;
```
success = (value != null) ? (boolean) value : false;

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
-    public final static int INDEX = 0, FILE = 1, TARGET = 2, TIMESTAMP_RECONSTRUCTION = 3, WAVELENGTH = 4, ALGORITHM = 5, RGL_WGT = 6, SUCCESS = 7, RATING = 8, COMMENTS = 9;
-    private static final String[] COLUMNS_NAMES = {"Index", "Name", "Target", "Timestamp reconstruction", "Wavelength", "Algorithm", "RGL_WGT", "Success", "Rating", "Comments"};
-
-    private List<ServiceResult> results;
```
add comment: copy of result list (really needed ?)

buthanoid replied:
If I understand the question correctly, the list of results seems needed to answer to `getValueAt`

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+    * when there is several columns with same name, only one is keeped based on priority.
+    * the decreasing priority is : OutputParam > InputParam > HardCoded.
+    */
+    private final LinkedHashSet<ColumnDesc> unionColumnDesc;
```
As discussed, just keep the complete List<ColumnDesc> (inventory computed by setResults); the userUnionColumnDesc should be handled in the view (basic table sorter) or controller (jtable has tableColumn(modelIndex)

buthanoid replied:
So we need to add Filter Columns capability into BasicTableSorter, because it does not have a model by itself, it decorates the model ResultSetTableModel.
I saw a method `computeColumnsIndirectionArray` in BasicTableSorter that computes a complete view of the columns. I suggest we modify it to be able to compute a different view which could filter some columns and re-order the displayed ones. What do you think ?

bourgesl replied:
Yes I am doing it today: 
1. I propose to add a new field List<String> viewColumns (null by default) to actually implement computeColumnsIndirectionArray() properly (very close to searchcal TableSorter).
2. TablePanel will have to deal with oimaging preferences (to get/set user preference)
=> the list of user columns will be handled in TablePanel (not model)

FYI BasicTableSorter seems quite outdated as swing JTable does provide sort + column filtering / reordering (see TableColumnModel) but it implies to refactor a lot of existing code ... (later or never)

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+    
+    /** originaly the same as unionColumnDesc but the GUI user can filter some columns. 
+     it is a list because then it is simpler to answer to columns index. */
+    private final List<ColumnDesc> userUnionColumnDesc;
```
keep only 1 list

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+        userUnionColumnDesc = new ArrayList<>();
     }
 
+    public Set<ColumnDesc> getUnionColumnDesc () {
```
should return List

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+        return Collections.unmodifiableSet(unionColumnDesc);
+    }
+    
+    /** read only of userUnionColumnDesc */
```
remove

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+        return Collections.unmodifiableList(userUnionColumnDesc);
+    }
+    
+    public void setUserUnionColumnDesc (List<ColumnDesc> userUnionColumnDesc) {
```
should go in tablePanel

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+        //    that there is a priority Output > Input > HardCoded on columns.
+        // 5. we reset userUnionColumnDesc to unionColumnDesc.
+        
+        List<ColumnDesc> listColumnDesc = new ArrayList<>();
```
As discussed, this algorithm can be simplied:
Set out, Set in
for (results) { out.add(outParams.keywords...) in.add(inParams.keywords...}
Now 2 sets containing union of outParams and inputParams
=> then build 1 global set with union of (output, input, hard coded)
=> build 1 final list (sorted or not): the view wil be handled in tablepanel (order)

For the transition (keywords = rating / comments, timestamps) => use a special source case (output or result) then getValueAt()/setValueAt() will try output then fallback to result ?

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756
use only one set but inserts Output then Input. I could change to the two sets method if it is considered clearer to the reader.
I kept the ordering for now. Surely it could be removed when we will implement the ordering in basicTableSorter.

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
     @Override
     public int getColumnCount() {
-        return COLUMNS_NAMES.length;
+        return userUnionColumnDesc.size();
```
this model should only deal with the list (unionColumnDesc)

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
-                return true;
-        }
-        return false;
+        ColumnDesc columnDesc = userUnionColumnDesc.get(columnIndex);
```
rowIndex should be checked to ensure result is editable: isValid() that means it can be edited

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+                    case COMMENTS: return result.getComments();
+                    default: 
+                        logger.warn("Unknown HardCodedColumnDesc :" + columnDesc);
+                        return null;
```
useless

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
                 if (result.getOifitsFile() != null) {
-                    return result.getOifitsFile().getWavelengthRange();
+                    final ImageOiInputParam param = result.getOifitsFile().getImageOiData().getInputParam();
+                    // Keywords
```
factorize this method in a static getKeywordValue(FitsTable table, key) that also test if the given table is null (error case)

buthanoid replied:
From what I see `getInputParam()` can never return `null`

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+                                return card.getValue();
+                            }
+                        }
+                        return null;
```
remove these useless return null

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+            case OUTPUT_PARAM:
                 if (result.getOifitsFile() != null) {
-                    return result.getOifitsFile().getImageOiData().getInputParam().getRglWgt();
+                    final ImageOiOutputParam param = result.getOifitsFile().getImageOiData().getOutputParam();
```
use getKeywordValue(FitsTable table, key)

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+                            }
+                        }
+                        return null;
+                    }
```
why 3 return null !

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+    }
+    
+    // please don't change the order as it is used as is, for sorting operations
+    public enum ColumnSource { HARD_CODED, INPUT_PARAM, OUTPUT_PARAM } 
```
To make ColumnDesc a generic class to put in jmcs, this enum is too specific => remove

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
 
-            case SUCCESS:
-                return result.isValid();
+    public static class ColumnDesc {
```
move this generic class in its own class file (I will move it in jmcs later)

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+    public static class ColumnDesc {
+        
+        /** source must be knowned to extract value for columns data. also used for ordering */
+        private ColumnSource source;
```
could be an int or String to stay generic and remain usable in switch(source) cases

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+        /** class of the elements in the columns */
+        private Class dataClass;
+        /** Prettier name for GUI display */
+        private String label;
```
this label is not used for now; maybe later the table header's renderer could be customized to use it ?

buthanoid replied:
In `BasicSetTableSorter` there is a class called `SortableHeaderRenderer` that is currently not used by now. This classes have prepared a tooltip but it misses the connection with `getDescription()` (which I recently added). It could surely be tuned to use` getLabel() `also.

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+            ColumnDesc other = (ColumnDesc) otherObject;
+            return getName().equals(other.getName());
+        }
+        @Override public int hashCode () { return Objects.hash(getName()); }
```
hashcode() needs a javadoc comment to explain it only uses name (for hashset use case): see setResults()

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+    /** Some HardCoded Columns. 
+     * It does not extends ColumnDesc because it already extends Enum */
+    public enum HardCodedColumn  {
+        INDEX(int.class, "Index"),
```
remove redundant fields (Target, wavelength, rgl_wgt) present in input/output

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +28,300 @@
+                switch (HardCodedColumn.valueOf(columnDesc.getName())) {
+                    case INDEX: return getRowCount() - rowIndex;
+                    case FILE: return result.getInputFile().getName();
+                    case TARGET:
```
remove redundant fields (Target, wavelength, rgl_wgt) present in input/output

bourgesl reviewed (commented):
Great work !

Minor changes needed, but in general:
- check nulls (oifits or param table can be null)
- fix imports / indentation (code formatting)

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/f01505045e671567ef2d9bf8c3456da4aa2fab5d

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+        jListAvailable.setModel(modelAvailable);
+        jListDisplayed.setModel(modelDisplayed);
+        
+        jListAvailable.updateUI();
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/62f5a272243c320976bcedfc6e21ca74f0316834

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+    }// </editor-fold>//GEN-END:initComponents
+
+    private void jButtonAddActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_jButtonAddActionPerformed
+        if (jListAvailable.getSelectedValue() != null) {
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/ca5ee5fa15502c60af81831c2c2a879842800330

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+        gridBagConstraints.gridy = 3;
+        add(jLabelDisplayedNb, gridBagConstraints);
+
+        jListAvailable.setSelectionMode(javax.swing.ListSelectionModel.SINGLE_SELECTION);
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/ca5ee5fa15502c60af81831c2c2a879842800330

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+    }//GEN-LAST:event_jButtonAddActionPerformed
+
+    private void jButtonRemoveActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_jButtonRemoveActionPerformed
+        if (jListDisplayed.getSelectedValue() != null) {
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/ca5ee5fa15502c60af81831c2c2a879842800330

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+    private final JDialog dialog;
+
+    // Constructor used when an edition has already been done
+    public TableEditorPanel(JDialog dialog, List<ResultSetTableModel.ColumnDesc> availableKeywords, List<ResultSetTableModel.ColumnDesc> keywordsDisplayed) {
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/50067c127a7ad20ecad37773cde8c9fe030cc2b8
the constructor does not modify its parameters anymore

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+        if (jListAvailable.getSelectedValue() != null) {
+            modelDisplayed.addElement(jListAvailable.getSelectedValue());
+            modelAvailable.removeElement(jListAvailable.getSelectedValue());
+            jLabelAvailableNb.setText(modelAvailable.getSize() + " available");
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/dca95f672cde22071d5c0ba162b14588e012a158

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+        if (jListDisplayed.getSelectedValue() != null) {
+            modelAvailable.addElement(jListDisplayed.getSelectedValue());
+            modelDisplayed.removeElement(jListDisplayed.getSelectedValue());
+        jLabelAvailableNb.setText(modelAvailable.getSize() + " available");
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/dca95f672cde22071d5c0ba162b14588e012a158

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+    }//GEN-LAST:event_jButtonRemoveActionPerformed
+
+    private void jButtonOkActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_jButtonOkActionPerformed
+        dialog.dispose();
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/fe9c7626169ba7ab58ee1e6c529c28408789e6f4

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+    // End of variables declaration//GEN-END:variables
+
+    @Override
+    public void mouseClicked(MouseEvent mouseEvent) {
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/be495a298eb16e6634fd18613dac426954210edb

bourgesl commented:
Please do less commits, maybe group fixes by classes so history & review is more obvious

bourgesl commented on src/main/java/fr/jmmc/oimaging/gui/TableEditorPanel.java:
```diff
@@ -0,0 +1,249 @@
+
+/**
+ * Panel to be added in a dialog box to edit the display of the table
+ * TODO FIX: The headers are cleared if the user presses the close button
```
Let's postpone the before/after actions to move a column among displayed columns...

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TablePanel.java:
```diff
@@ -24,11 +30,19 @@
     private final BasicTableSorter resultSetTableSorter;
+    
+    /** object with rendering and editing responsabilities for column RATING.
+     * Careful ! You must have one RatingCell for each RATING column.
```
indeed, since I saw the field `private int rating;` I assumed that it was only meant for one column. I guess this field is used only sequentially by columns.

buthanoid replied:
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

bourgesl commented on src/main/java/fr/jmmc/oimaging/model/ResultSetTableModel.java:
```diff
@@ -16,126 +29,304 @@
+            if (result.getOifitsFile() == null) continue;
+            ImageOiOutputParam output = result.getOifitsFile().getImageOiData().getOutputParam();
+            for (KeywordMeta keyMeta : output.getKeywordsDesc().values()) {
+                setColumnDesc.add(new ColumnDesc(ColumnSource.OUTPUT_PARAM,keyMeta.getName(), keyMeta.getClass())); 
```
keyMeta.getClass() or card.getClass() do not give the data type !
Please use Types.getBaseClass(keyMeta.getDataType()) for keywords.

TODO Laurent: I will write a new Object parseValue() in header card to properly return null or String/Integer/Double/Boolean values according to the internal String value

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TablePanel.java:
```diff
@@ -76,14 +87,68 @@ private void initComponents() {
+        dialog.setVisible(true);
+        
+        // when dialog returns we set the chosen columns by user
+        setUserUnionColumnDesc(tableEditorPanel.getKeywordsToDisplay());
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/fe9c7626169ba7ab58ee1e6c529c28408789e6f4

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TablePanel.java:
```diff
@@ -37,20 +51,17 @@ public TablePanel() {
+
         jResultSetTable.setModel(resultSetTableSorter);
         SwingUtils.adjustRowHeight(jResultSetTable);
-
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TablePanel.java:
```diff
@@ -76,14 +87,68 @@ private void initComponents() {
+            final TableColumn columnRating = jResultSetTable.getColumn(ResultSetTableModel.HardCodedColumn.RATING.toString());
+            columnRating.setCellRenderer(ratingCell);
+            columnRating.setCellEditor(ratingCell);
+        } catch (IllegalArgumentException e) {}
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

buthanoid commented on src/main/java/fr/jmmc/oimaging/gui/TablePanel.java:
```diff
@@ -76,14 +87,68 @@ private void initComponents() {
+     * @param userUnionColumnDesc 
+     */
+    public void setUserUnionColumnDesc(List<ResultSetTableModel.ColumnDesc> userUnionColumnDesc) {
+        getTableModel().setUserUnionColumnDesc(userUnionColumnDesc);
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756
unfinished (removed from ResultSetTableModel but functionnality not restablished yet)

buthanoid commented on src/main/java/fr/jmmc/oimaging/model/RatingCell.java:
```diff
@@ -72,7 +72,7 @@ public Component getTableCellEditorComponent(JTable table, Object value, boolean
     public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
         rating = UNDEFINED;
-        starRater.setSelection((int) value);
+        if (value != null) starRater.setSelection((int) value);
```
https://github.com/JMMC-OpenDev/oimaging/pull/3/commits/85eacbb838ac180bcac14f40de59fd871bd32756

buthanoid commented:
Still misses :
- HeaderCard.getBaseClass() 
- BasicTableSorter view

bourgesl commented:
Please have a look and test my patch on your PR branch.
To make BasicTableSorter / user preferences work, I always use List<String> columns.
I wonder if the TableEditor should be based on List<String> too to become more generic, but also handle missing columns ! See the TablePanel methods that handle conversions between List<ColumnDesc> <=> List<String> ...

For example, the user preference could contain more column names (obtained on MIRA + BSMEM ...) but opening the editor (at startup or when results are coming from other software) will loose the unknown columns (as ColumnDesc are missing) ...

Finally handling global dynamic inventory of potential keywords / header cards is really tricky...

buthanoid commented:
What if instead of registering preferences for visible columns, we register preferences for explicitly hidden columns ?
Currently, the following scenario : 
1. Launch OImaging and Run Wisard
2. Explicitly hide only INDEX column
3. Close OImaging
4. Launch OIMaging and Run BSMEM
will hide not only INDEX but also all BSMEM columns that were not part of WISARD.
If we register preferences for hidden columns, the scenario would then only hide INDEX.

bourgesl commented:
Antoine, I fixed my previous comment to explain a TableEditor problem with  `List<ColumnDesc>` <=> `List<String>`
Your idea about hidden columns is interesting, but the explicit visible columns helps to have an ordered list of columns too.
I think the problem has two aspects:
- preserve unknown columns in the table editor (and maintain their position in the list of visible columns): this should be handled in the TableEditor by using column names as String and preserving column names not present in the current allColumns
- always show new columns (not present in the previous list of all columns)

I propose to introduce a new preference to store all column names (as string) to have 2 preferences with (all, ordered + visible) column names to determine easily new columns (all columns - pref's all columns) => show them at the end ?

buthanoid commented:
If I understood well, this is the current : 
- the _list of current results columns_, field of ResultSetTableModel, updated every time the service results list changes.
- the _list of visible columns_, a home preference (persistent between JVM runs), updated every time the user drag&drops a column (order change), every time the user complete the TableEditorPanel dialog (it copies the resulting list).
- the _list of displayed columns_, a field of BasicTableSorter, which is always equals to the _list of visible columns_.
- the _list of hidden columns_, a field of TableEditorPanel, which is initialized to the _list of current results columns_, minus the _list of visible columns_.
- _another list of displayed columns_, a field of TableEditorPanel, which is initialized to the _list of current results columns, intersected with the _list of visible columns_.

It presents a problem : 
When we add a new service result, any new column is well added to the _list of current results columns_, but is not displayed because it does not belong to the _list of visible columns_. If we don't display the column, it is not ergonomic because the user may wonder why its columns does not show whereas he just ran the algorithm. I we display the column, it can trouble the user because every time he always removes this column from the visible ones, but he has to do this every time he runs the OImaging software, because it is forgetting the setting. Only visible columns are remembered, not hidden columns. Here we are missing some information to address the issue.

You propose to add another list : 
- the _list of all encountered columns_, a home preference, which remembers every column ever saw. 
And so, to make the following changes : 
- When a result is added, and one of its columns does not belong to the _list of all encountered columns_, we add the column to that list, and also to the _list of visible columns_.
- When a result is added, and one of its columns belongs to the _list of all encountered columns_, we have nothing to do : either it belongs to the _list of visible columns_, either the user explicitly removed it from.

Is this correct ?

I see another problem : 
In TableEditorPanel we only sublists of the _list of current results columns_. It seems good for GUI to not trouble the user with columns that are not even part of its results.
Problem is, when the user completes the dialog, we set the _list of visible columns_ (home preference). Thus, we loose any information about other columns. Even with the solution of the _list of all encountered columns_, we still loose the visible and the order in the _list of visible columns_.

bourgesl commented:
Very good summary of the current situation !

Concerning your last point, I think the table editor should be String based to let display any given column name (real or virtual) in the displayed columns = user preference = BasicTableSorter visible column names,  even if this column does not belong to the current _list of current results columns_.
I agree GUI should be wysiwyg, but in this case, the preference indicated this column was already present / seen by the user and is coming from previous sessions...

I think the round-trip (preference -> table editor -> no change -> OK) should be consistent and stable ie not loose any visible column.
Finally it looks really tricky and very complicated.
Any other idea to solve this algebra problem ?

buthanoid commented:
To try to address the problem : 

- we have the set of all encountered columns. stored in home preference. we will name it ALL.
- we have the list of visible columns, an ordered subset of ALL. stored in home preference. we will name it ALL_VISIBLE.
- we have the set of union of current results columns, a subset of ALL (we complete ALL if necessary). stored in a field of  ResultSetTableModel (as a list for technical issues, but here we don't care about its order). we will name it RESULTS.

Then : 
- we have the list (ordered set) of displayed columns in the GUI, which ALL_VISIBLE, intersected by RESULTS. stored in a field of BasicTableSorter. We will name it VISIBLE_RESULTS.
- we have the set of hidden columns in the GUI, which is RESULTS minus VISIBLE_RESULTS. created for each TableEditorPanel. we will name it HIDDEN_RESULTS. This is the list of avalaible keywords in TableEditorPanel.
- This list of displayed keywords in TableEditorPanel is taken from VISIBLE_RESULTS.

The process :

- when a result is added:
    - for every column that is not in ALL :
        - add the column to ALL
        - add the column to the end of ALL_VISIBLE
    - recompute RESULTS (union of columns, function already exists)
    - recompute VISIBLE_RESULTS (intersection, easy to program)
    
- when a result is removed:
    - recompute RESULTS
    - recompute VISIBLE_RESULTS
    
- when the user start TableEditorPanel:
    - copy VISIBLE_RESULTS for displayedKeywords field
    - compute HIDDEN_RESULTS (minus, easy to program) for avalaibleKeywords field
    
- when the user completes TableEditorPanel with success:
    - update ALL_VISIBLE to f(ALL_VISIBLE, resultOfTableEditorPanel). f is detailed below
    - set VISIBLE_RESULTS to resultOfTableEditorPanel
    
- when a column is drag & dropped:
    - update ALL_VISIBLE to f(ALL_VISIBLE, VISIBLE_RESULTS)
    - no need to recompute VISIBLE_RESULTS, it is already correct
   
   
   
The function f takes a big list, and a sublist of the big list that have been re-ordered. It applies the re-ordering to the big list.

Pseudo code :
```

// requirements : bigList and subList, while ordered, dont have duplicates
//                subList is a subset of bigList

f (bigList, subList, i, j) = {

   int j = 0;

   for (i = 0 ; i < bigList.length ; i ++) {
   
      if (bigList[i] belongs to subList) {
         bigList[i] = subList[j];
         j ++ ;
      }
   }

}
```

bourgesl commented:
That specs looks a good case study !
I wonder what final order is given by f() once the table editor is done
It implies table editor will be string based, in these steps we will have to handle both list of column desc & list of column names, but it seems doable.
Go ?

bourgesl commented:
As BasicTableSorter ignores column names not present in the data model, it is not necessary to compute `VISIBLE_RESULTS`. 
It may simplify the overall algorithm, it only impacts the table editor constructor.

buthanoid commented:
Good point.
I am reviewing what I wrote yesterday. For example f is incomplete because it only address order issues whereas there is also a change of columns.

buthanoid commented:
Data:
- `ALL` the set of all columns met. home preference, persistent.
- `ALL_VISIBLE` the list of all columns met that the users want to display, in that order. home preference, persistent.
- `RESULTS` the list, union of all columns of all current results. field of `ResultSetTableModel`. The order is important to query the data, but note that this is not the order used for GUI display, for that see `BasicTableSorter`.
- `VISIBLE_RESULTS` the columns actually displayed by `BasicTableSorter`. No need to compute it ourselves : `BasicTableSorter` will use `RESULTS` as data, and we set `BasicTableSorter.visibleColumnNames` to `ALL_VISIBLE`. The class will use the order of `ALL_VISIBLE`, but will manage the intersection with `RESULTS`, and the correct data mapping to `RESULTS`.
- `HIDDEN_COLUMNS` the list of hidden columns in `TablePanelEditor`. It must only contains columns that belong to current results. see computing below.
- `DISPLAYED_COLUMNS` the list of displayed columns in `TablePanelEditor`.  It must only contains columns that belong to current results. see computing below.
- `NEW_ALL_VISIBLE` the temporary list of `ALL_VISIBLE` when user uses `TablePanelEditor`. it is simpler to modify this list each time the user activates or hides a columns, than to compute it afterwards. see computing below.

Process:

- when a result is added:
   - for each column not in `ALL`:
      - add the column to `ALL`
      - add the column to the end of `ALL_VISIBLE` (a new met column must be displayed)
   - recompute `RESULTS` (existing function `ResultSetTableModel.setResults`)
   - set `BasicTableSorter.visibleColumnNames` to `ALL_VISIBLE`

- when a result is removed:
   - recompute `RESULTS`

- when the user starts `TableEditorPanel`:
   - compute `NEW_ALL_VISIBLE`: 
      - copy `ALL_VISIBLE`
   - compute `DISPLAYED_COLUMNS`:
      - copy `ALL_VISIBLE`
      - intersect it with `RESULTS` (note that then we keep the order of `ALL_VISIBLE`)
   - compute `HIDDEN_COLUMNS`:
      - copy `RESULTS`
      - remove `DISPLAYED_COLUMNS` from it
      
- when the user activates a column on `TableEditorPanel`:
   - add the column to the end of `NEW_ALL_VISIBLE`
   - add the column to the end of `DISPLAYED_COLUMNS`
   - remove the column from `HIDDEN_COLUMNS`
   
- when the user hides a column on `TableEditorPanel`:
   - add the column to the end of `HIDDEN_COLUMNS`
   - remove the column from `DISPLAYED_COLUMNS`
   - remove the column from `NEW_ALL_VISIBLE`

- when the user completes `TableEditorPanel` with success:
   - set `ALL_VISIBLE` to `NEW_ALL_VISIBLE`
   - set `BasicTableSorter.visibleColumnNames` to `ALL_VISIBLE`
   - discard `NEW_ALL_VISIBLE`, `HIDDEN_COLUMNS` and `DISPLAYED_COLUMNS`
   
- when a column is drag & dropped:
   - compute `NEW_DISPLAYED_COLUMNS`:
      - copy `BasicTableSorter.visibleColumnNames`
   - call `reorderFromSubList(ALL_VISIBLE, NEW_DISPLAYED_COLUMNS)`
   
   
```c
// reorderFromSubList reorder a list, by following the order of given subList :
// - only elements present in subList are moved in the list
// - the others elements do not move
   
// requirements : list do not contain duplicates
//                subList do not contain duplicates
//                subList is a subset of list 

reorderFromSubList (list, subList) = {

   int j = 0;

   for (i = 0 ; i < list.length ; i ++) {
   
      if (list[i] belongs to subList) {
         list[i] = subList[j];
         j ++ ;
      }
   }
}
```
Note that `reorderFromSubList` is used only in drag&drop. but if we implement one day "up" and "down" buttons in `TableEditorPanel`, we will need it there too.

buthanoid commented:
I think we do not need `reorderFromSubList` since `BasicTableSorter.getVisibleColumnNames` already returns the complete list (not just the ones present in results)

bourgesl commented:
Drag n drop in table will change visible column names to only visible & existing columns.

To simplify this overly too complicated problem due to non standard keywords (header cards), I think unknown columns must be displayed even empty = not present in the current results.
It means unknown columns must be added in the table model: new ColumnDesc(name, undefined source) then basic sorter / table editor will display them... and no more ambiguous / unstable behaviour.

bourgesl commented:
Here is my alternative: https://github.com/bourgesl/oimaging/tree/dynamic-table-simple
It seems stable (always columns are displayed), I can later auto-hide empty columns...
(sorry it includes also your PR#3 + few changes from your other branch) + needs jmcs (ColumnDesc)

bourgesl commented:
I am merging again...

bourgesl commented:
Beta release: http://jmmc.fr/~betaswmgr/OImaging/
